src/gc.c

```diff
@@ -3272,6 +3278,44 @@ static int _jl_gc_collect(jl_ptls_t ptls, jl_gc_collection_t collection)
         gc_num.last_full_sweep = gc_end_time;
     }
 
+    size_t heap_size = jl_atomic_load_relaxed(&gc_heap_stats.heap_size);
+    uint64_t alloc_diff = before_free_heap_size - old_heap_size;
+    uint64_t freed_diff = before_free_heap_size - heap_size;
+
+    double alloc_smooth_factor = 0.95;
+    double collec_smooth_factor = 0.5;
+    double tuning_factor = 0.03;
+    double alloc_mem = jl_gc_smooth(old_alloc_diff, alloc_diff, alloc_smooth_factor);
+    double alloc_time = jl_gc_smooth(old_mut_time, mutator_time, alloc_smooth_factor);
+    double gc_mem = jl_gc_smooth(old_freed_diff, freed_diff, collec_smooth_factor);
+    double gc_time = jl_gc_smooth(old_pause_time, pause, collec_smooth_factor);
+    old_alloc_diff = alloc_diff;
+    old_mut_time = mutator_time;
+    old_freed_diff = freed_diff;
+    old_pause_time = pause;
+    old_heap_size = heap_size;
+    double min_interval = default_collect_interval;
+    double target_allocs;
+    if (alloc_mem != 0 && alloc_time != 0 && gc_mem != 0 && gc_time != 0) {
+        double alloc_rate = alloc_mem/alloc_time;
+        double gc_rate = gc_mem/gc_time;
+        target_allocs = sqrt(((double)heap_size/min_interval * alloc_rate)/(gc_rate * tuning_factor)); // work on multiples of min interval
+    }
+    else
+        target_allocs = 2*sqrt((double)heap_size/min_interval);
+
+    uint64_t target_heap = (uint64_t)target_allocs*min_interval + heap_size;
+    if (target_heap > max_total_memory)
+        target_heap = max_total_memory;
+    else if (target_heap < default_collect_interval)
+        target_heap = default_collect_interval;
+    jl_atomic_store_relaxed(&gc_heap_stats.heap_target, target_heap);
```

Comment:
We should probably keep this behind a guard
```
if (collection == JL_GC_AUTO) {
```
Such that user calls to GC.gc don't end up causing the 0.95% smoothing factor to be applied over too short of an interval, causing the user to blow up their GC with small numbers.